force_wfmanager/ui/setup/mco/tests/test_kpi_specification_view.py

```diff
@@ -0,0 +1,144 @@
+    KPISpecificationView, KPISpecificationModelView, TableRow
+
+
+class TestKPISpecificationModelView(unittest.TestCase, UnittestTools):
+
+    def setUp(self):
+
+        self.kpi_model_view = KPISpecificationModelView(
+            model=KPISpecification(name='kpi_test')
+        )
+
+    def test_kpi_model_view_init(self):
+        self.assertEqual(
+            "KPI: kpi_test (MINIMISE)",
+            self.kpi_model_view.label
+        )
+        self.assertTrue(self.kpi_model_view.valid)
+
+    def test_kpi_label(self):
+        self.kpi_model_view.model.name = 'kpi_name_test'
+        self.assertEqual(
+            "KPI: kpi_name_test (MINIMISE)",
+            self.kpi_model_view.label
+        )
+
+    def test_verify_workflow_event(self):
+        with self.assertTraitChanges(
+                self.kpi_model_view, 'verify_workflow_event', count=1):
+            self.kpi_model_view.model.name = 'another'
+
+
+class TestKPISpecificationView(unittest.TestCase, UnittestTools):
+
+    def setUp(self):
+        self.registry = get_basic_variable_names_registry()
+        self.workflow = self.registry.workflow
+        self.param1 = self.workflow.mco.parameters[0]
+        self.param2 = self.workflow.mco.parameters[1]
+        self.param3 = self.workflow.mco.parameters[2]
+        self.data_source1 = self.workflow.execution_layers[0].data_sources[0]
+        self.data_source2 = self.workflow.execution_layers[0].data_sources[1]
+        self.workflow.mco.kpis.append(KPISpecification())
+
+        self.kpi_view = KPISpecificationView(
+            model=self.workflow.mco,
+            variable_names_registry=self.registry
+        )
+
+    def test_kpi_view_init(self):
+        self.assertEqual(1, len(self.kpi_view.kpi_names))
+        self.assertEqual(1, len(self.kpi_view.kpi_model_views))
+        self.assertEqual('KPI', self.kpi_view.kpi_model_views[0].label)
+        self.assertEqual("MCO KPIs", self.kpi_view.label)
+
+    def test_label(self):
+        self.data_source1.output_slot_info = [OutputSlotInfo(name='T1')]
+
+        self.assertEqual(['T1'], self.registry.data_source_outputs)
+
+        self.assertEqual('KPI', self.kpi_view.kpi_model_views[0].label)
+
+        self.workflow.mco.kpis[0].name = 'T1'
+
+        self.assertEqual(
+            "KPI: T1 (MINIMISE)",
+            self.kpi_view.kpi_model_views[0].label
+        )
+
+        self.workflow.mco.kpis[0].objective = 'MAXIMISE'
+
+        self.assertEqual(
+            "KPI: T1 (MAXIMISE)",
+            self.kpi_view.kpi_model_views[0].label
+        )
+
+    def test_name_change(self):
+
+        self.data_source1.output_slot_info = [OutputSlotInfo(name='T1')]
+        kpi_model_view = self.kpi_view.kpi_model_views[0]
+
+        with self.assertTraitChanges(kpi_model_view, 'label', count=0):
+            self.assertEqual('KPI', kpi_model_view.label)
+
+        with self.assertTraitChanges(kpi_model_view, 'label', count=1):
+            self.workflow.mco.kpis[0].name = 'T1'
```

Comment:
As commented in the corresponding module, we should define a behaviour for when the output_slot changes in the data source